### PR TITLE
hyperHTML.adopt(node)

### PR DESCRIPTION
**Pull request by WebReflection**
As mentioned in the closed [issue 38](https://github.com/WebReflection/hyperHTML/issues/38#issuecomment-296129937), there are cases where `hyperHTML` might be used to adopt an already existing node, instead of trashing its content at the very first pass.

### Experimental
The `hyperHTML.adopt(node)` is an alternative to `hyperHTML.bind(node)` and it's designed to match "_best effort_" a live node content against a template literals.

Ideally, the live node content should've been created via [viperHTML](https://github.com/WebReflection/viperHTML) to be sure the mapping between nodes, content, and attributes, is exactly the expected one.

#### Note about wires
I'd like to first release this `.adopt(node)` method and see how it goes first.
I will eventually figure out how to adopt wires too but right now, every wire, will still create once the wired node instead of adopting it (also adopt is about the node content, not the node itself).

TL;DR wires work just like before, if you adopt a node full of wires you'll still trash them at first pass.

### CODE REVIEW

File: hyperhtml.js [hunk 1]
```diff
@@ -278,15 +305,61 @@ var hyperHTML = (function () {'use strict';
...
-      i = 0,
-      length = childNodes.length;
-      i < length; i++
-    ) {
-      node.appendChild(childNodes[i]);
+  // given a generic text or comment node
+  // remove the surrounding emptiness
+  function cleanAround(textNode, sibling) {
+    var adj = textNode[sibling];
+    if (adj && adj.nodeType === 3 && trim.call(adj.textContent).length < 1) {
```
Comment (marcoscaceres) [inline hyperhtml.js]: Nit: if it's all the same, and assuming inlining by the JIT, s/3/Node.TEXT_NODE ... just makes it a little more readable.

File: hyperhtml.js [hunk 2]
```diff
@@ -278,15 +305,61 @@ var hyperHTML = (function () {'use strict';
...
+      textNode.parentNode.removeChild(adj);
+      cleanAround(textNode, sibling);
+    }
+  }
+
+  // find a live node through a virtual one
+  function find(live, virtual) {
+    var i, length, name, parentNode, map = [];
+    switch(virtual.nodeType) {
+      case 1:
```
Comment (marcoscaceres) [inline hyperhtml.js]: Nit: I guess the same as above here - or is there some better reason you are just using numbers here?

File: hyperhtml.js [hunk 3]
```diff
@@ -278,15 +305,61 @@ var hyperHTML = (function () {'use strict';
...
+    var adj = textNode[sibling];
+    if (adj && adj.nodeType === 3 && trim.call(adj.textContent).length < 1) {
+      textNode.parentNode.removeChild(adj);
+      cleanAround(textNode, sibling);
+    }
+  }
+
+  // find a live node through a virtual one
+  function find(live, virtual) {
+    var i, length, name, parentNode, map = [];
```
Comment (marcoscaceres) [inline hyperhtml.js]: nit: any reason you switched to c style variable declarations here? (i.e., why not declare i, length, in the for loop?

File: hyperhtml.js (hunk 1 above)
Comment (WebReflection) [inline hyperhtml.js]: I need a global Node first to do that ... remember this project is used also in [nativeHTML](https://github.com/WebReflection/nativeHTML) through [basicHTML](https://github.com/WebReflection/basicHTML) and the less I assume globals the better it is.

It's already tricky to initialize `hyperHTML` in `basicHTML` due one off feature detection using `document`: https://github.com/WebReflection/basicHTML/blob/master/test/hyperhtml.js

File: hyperhtml.js (hunk 3 above)
Comment (WebReflection) [inline hyperhtml.js]: in case you haven't noticed, there is never more than a single `var` declaration per function so nobody can ever possibly be confused by the hoisting nature of the `var`.

I'd rather go ES6 when the time is right. 'till that day, I'd like to have zero tooling dependencies, hoping whoever uses this lib is also not transpiling much more than just  what `babel-plugin-transform-es2015-template-literals` would do :wink:

File: hyperhtml.js (hunk 2 above)
Comment (WebReflection) [inline hyperhtml.js]: same as above, although I do appreciate your reviews and nitpicks and I'll try to wait for them next PR (I have to remember to ask you to review via GitHub, sorry I didn't this time)

Comment (WebReflection) [inline hyperhtml.js]: about numbers though, I might create private shared variables with these numbers in without needing the Node global ... actually, I'm doing this now :laughing:
